core/logging.py

```diff
@@ -0,0 +1,144 @@
+    """
+    Manages PLAMS logger instances.
+    The manager should not be instantiated directly, but loggers accessed through the ``get_logger`` method.
+    """
+
+    _loggers: Dict[str, "Logger"] = {}
+
+    def __new__(cls, *args, **kwargs):
+        raise TypeError("LoggerManager cannot be directly instantiated.")
+
+    @classmethod
+    def get_logger(cls, name: str) -> "Logger":
+        """
+        Get a logger with the specified name.
+        If there is an existing logger with the same name this is returned, otherwise a new logger is created.
+        """
+        if name not in cls._loggers:
+            cls._loggers[name] = Logger(name)
+        return cls._loggers[name]
+
+
+class Logger:
+    """
+    Wrapper around default logger, which handles logging to stdout and a text logfile depending on the options in the
+    global logging config.
+    """
+
+    def __init__(self, name: str):
+        """
+        Get a logger with given name.
+        """
+        self._logger = logging.Logger(name)
+        self._stdout_handler: Optional[logging.StreamHandler] = None
+        self._file_handler: Optional[logging.FileHandler] = None
+        self._formatter: Optional[logging.Formatter] = None
+        self._lock = threading.Lock()
+
+    def configure(
+        self,
+        stdout_level: int = 0,
+        logfile_level: int = 0,
+        logfile_path: Optional[str] = None,
+        include_date: bool = False,
+        include_time: bool = False,
+    ) -> None:
+        """
+        Configure logging to stdout and the logfile, and its formatting.
+
+        For backwards compatibility, the logging level is between 0-7, with 0 indicating no logging and 7 the most verbose logging.
+        Note that this is only a PLAMS logging level, it will be mapped to a level between INFO and WARNING for the
+        standard python logger.
+
+        The logfile path must be unique across all loggers, otherwise a |FileError| is raised. If set to None this will disable file logging.
+
+        :param stdout_level: value between 0-7, with 0 indicating no logging and 7 the most verbose logging to stdout
+        :param logfile_level: value between 0-7, with 0 indicating no logging and 7 the most verbose logging to logfile
+        :param logfile_path: path for the logfile, if set to None this will remove file logging
+        :param include_date: whether to include date stamp at the start of a log line
+        :param include_time: whether to include time stamp at the start of a log line
+        """
+        with self._lock:
+
+            # Initialise the stdout handler once
+            if self._stdout_handler is None:
+                self._stdout_handler = logging.StreamHandler(sys.stdout)
+                self._stdout_handler.setFormatter(self._formatter)
+                self._logger.addHandler(self._stdout_handler)
+
+            # Update the stdout handler level if required
+            if stdout_level != self._stdout_handler.level:
```

Comment:
I would suggets it should be:

`if stdout_level != 28 - self._stdout_handler.level:`

Reply:
Thanks, good spot - updated